modules/data.atmosphere/R/extract_local_CMIP5.R

```diff
@@ -0,0 +1,299 @@
+    } # End file loop
+  } # End variable loop
+    
+  # Dealing with leap-year post-hoc because it was becoming a pain in the ass
+  # If we have daily data and we're dealing with a model that skips leap year, add it in
+  dpm <- days_in_month(1:12)
+  yrs.leap <- ylist[leap_year(ylist)]
+  for(y.now in yrs.leap){
+    yr.ind <- which(year(dat.time)==y.now)
+    if(GCM %in% no.leap & v.res == "day"){
+      for(v in 1:length(dat.all)){
+        dat.all[[v]] <- append(dat.all[[v]], dat.all[[v]][yr.ind[sum(dpm[1:2])]], sum(yr.ind[dpm[1:2]]))
+      }
+    }
+  }
+  
+  
+
+  print("")
+  print("- Writing to NetCDF: ")
+  pb <- txtProgressBar(min=1, max=rows, style=3)
+  for (i in 1:rows){
+    setTxtProgressBar(pb, i)
+    
+    y.now = ylist[i]    
+    yr.ind <- which(year(dat.time)==y.now)
+    
+    
+    dpm <- days_in_month(1:12)
+    if(leap_year(y.now)) dpm[2] <- dpm[2] + 1 # make sure Feb has 29 days if we're dealing with a leap year
+    
+    # figure out how many days we're working with
+    if(rows>1 & i!=1 & i!=rows){ # If we have multiple years and we're not in the first or last year, we're taking a whole year
+      nday  = ifelse(lubridate:: leap_year(y.now), 366, 365) # leap year or not; days per year
+      day1 = 1
+      day2 = nday
+      days.use = day1:day2
+    } else if(rows==1){
+      # if we're working with only 1 year, lets only pull what we need to
+      nday  = ifelse(lubridate:: leap_year(y.now), 366, 365) # leap year or not; days per year
+      day1 <- yday(start_date)
+      # Now we need to check whether we're ending on the right day
+      day2 <- yday(end_date)
+      days.use = day1:day2
+      nday=length(days.use) # Update nday
+    } else if(i==1) {
+      # If this is the first of many years, we only need to worry about the start date
+      nday  = ifelse(lubridate:: leap_year(y.now), 366, 365) # leap year or not; days per year
+      day1 <- yday(start_date)
+      day2 = nday
+      days.use = day1:day2
+      nday=length(days.use) # Update nday
+    } else if(i==rows) {
+      # If this is the last of many years, we only need to worry about the start date
+      nday  = ifelse(lubridate:: leap_year(y.now), 366, 365) # leap year or not; days per year
+      day1 = 1
+      day2 <- yday(end_date)
+      days.use = day1:day2
+      nday=length(days.use) # Update nday
+    }
+    ntime = nday # leap year or not; time slice (coerce to daily)
+    
+    loc.file <- file.path(outfolder, paste(model, scenario, ensemble_member, str_pad(y.now, width=4, side="left",  pad="0"), "nc", sep = "."))
+    
+    
+    ## Create dimensions
+    dim.lat <- ncdim_def(name='latitude', units='degree_north', vals=lat.in, create_dimvar=TRUE)
+    dim.lon <- ncdim_def(name='longitude', units='degree_east', vals=lon.in, create_dimvar=TRUE)
+    dim.time <- ncdim_def(name='time', units="sec", vals=seq((min(days.use)+1-1/24)*24*360, (max(days.use)+1-1/24)*24*360, length.out=ntime), create_dimvar=TRUE, unlim=TRUE)
```

Comment:
Not sure what CF time units are for historical GCMs, but future GCMs time should be "days since YYYY".  Just so we are keeping consistent with the direction we want to move in for dimensions, I think we should consider editing this. Here is an example from download.CRUNCEP.

days_elapsed <- (1:ntime) * 6/24 - 3/24 # data are 6-hourly, with timestamp at center of interval
time <- ncdf4::ncdim_def(name = "time", units = paste0("days since ", year, "-01-01T00:00:00Z"),
                             vals = as.array(days_elapsed), create_dimvar = TRUE, unlim = TRUE)

Reply:
I think I found the past GCMs to be inconsistent and wasn't sure if PEcAn had a standard reference time that would make sense for our p1000 dates.